Source/RunActivity/Viewer3D/RollingStock/MSTSLocomotiveViewer.cs

```diff
@@ -2174,8 +2155,8 @@ public virtual int GetDrawIndex()
                         index = PercentToIndex(Locomotive.GetCombinedHandleValue(false));
                     break;
                 case CABViewControlTypes.CP_HANDLE:
-                    if (Locomotive.CombinedControlType == MSTSLocomotive.CombinedControl.ThrottleDynamic && Locomotive.DynamicBrakePercent >= 0
-                            || Locomotive.CombinedControlType == MSTSLocomotive.CombinedControl.ThrottleAir && Locomotive.TrainBrakeController.CurrentValue > 0)
+                    if (Locomotive.CombinedControlType == MSTSLocomotive.CombinedControl.ThrottleDynamic && Locomotive.DynamicBrakeController?.CurrentValue > 0
+                            || Locomotive.CombinedControlType == MSTSLocomotive.CombinedControl.ThrottleAir && Locomotive.TrainBrakeController?.CurrentValue > 0)
                             index = PercentToIndex(Locomotive.GetCombinedHandleValue(false));
                         else
                             index = PercentToIndex(Locomotive.GetCombinedHandleValue(false));
```

Comment:
```suggestion
                    index = PercentToIndex(Locomotive.GetCombinedHandleValue(false));
```
Probably we should remove the if clause. The body is the same for both branches.

Reply:
Weird. This section is bothering me the more I look at it. Looks like this might have been set up to handle the case where dynamic brakes were activated but at 0%, but that's entirely pointless.

Then there's the whole `index = (int)((ControlDiscrete.FramesCount) * Locomotive.GetCombinedHandleValue(false))` case. CJ's comment justifying this implementation talks about 11.11% instead of 12.5%, but I don't think that's the problem here (or, maybe that was a problem but the changes to dynamic brakes in this PR fixed it). I found that if I unify all cases to use `index = PercentToIndex(Locomotive.GetCombinedHandleValue(false))` the _only_ problem with the MSTS Dash 9 is that dynamic braking notch 3 shows "B2" on the combined power display. And this is sort of a bug in the Dash 9 cab view, as the `SwitchVal` for the B3 frame is set to 0.73, when it should be 0.72222 (the actual combined control value when in B3).

Of course, it worked fine in MSTS, which implies `PercentToIndex` is not functioning the same as the equivalent code in MSTS. Looking at some MSTS videos, I have to guess that (in braking at least), MSTS would look for the index associated with the smallest value _greater_ than the percent, while the ORTS implementation is specifically looking for the largest value _below_ the given percent regardless of power or braking. Neither way of doing things is all that great, I'd like to try just matching for the closest index, higher or lower, and seeing what happens.

I'm going to unify the CPH cab view calculations and give reworking PercentToIndex a shot. It's a bit of a bold move, but I don't think it's too risky as the difference will be minor and difficult to notice. Like, the current implementation would show "B3" on the Dash 9 for an input of 0.73 to 0.77775, but a "closest index" implementation would show "B3" between 0.698 to 0.754, sliding back the range by just a few percent (and including 0.72222 as intended).

Reply:
Someone reported that indicators stopped working after the refactoring
```
CombinedControl (
Type ( CP_HANDLE COMBINED_CONTROL )
Position ( 426.875 397.03125 176.25 82.96875 )
Graphic ( "..\\..\\1CABVIEW\\S100\\CD_CP.ace" )
Style ( NONE )
MouseControl ( 1 )
NumFrames ( 3 3 1 )
NumPositions ( 3 0 1 2 )
NumValues ( 3 1 0.5 0 )
Orientation ( 1 )
DirIncrease ( 0 )
ScaleRange ( 0 1 )
)
Lever (
Type ( DYNAMIC_BRAKE LEVER )
Position ( 426.875 397.03125 176.25 82.96875 )
Graphic ( "..\\..\\1CABVIEW\\S100\\CD_dec.ACE" )
Style ( SPRUNG )
MouseControl ( 0 )
NumFrames ( 2 2 1 )
NumPositions ( 2 0 1  )
NumValues ( 2 0 0.01 )
Orientation ( 1 )
DirIncrease ( 1 )
ScaleRange ( 0 0.01 )
)
CombinedControl (
Type ( CP_HANDLE COMBINED_CONTROL )
Position ( 262.5 459.375 138.125 15.46875 )
Graphic ( "..\\..\\1CABVIEW\\S100\\flecha_intensidadPuff.ace" )
Style ( NONE )
MouseControl ( 1 )
NumFrames ( 54 2 27 )
NumPositions ( 0 )
NumValues ( 0 )
Orientation ( 1 )
DirIncrease ( 0 )
ScaleRange ( 0 1 )
)
CombinedControl (
Type ( CP_HANDLE COMBINED_CONTROL )
Position ( 169.375 265.78125 10 11.25 )
Graphic ( "..\\..\\1CABVIEW\\S100\\ind_inv.ace" )
Style ( NONE )
MouseControl ( 1 )
NumFrames ( 3 3 1 )
NumPositions ( 3 0 1 2 )
NumValues ( 3 1 0.5 0 )
Orientation ( 1 )
DirIncrease ( 0 )
ScaleRange ( 0 1 )
)
Lever (
Type ( DYNAMIC_BRAKE LEVER )
Position ( 169.375 265.78125 10 11.25 )
Graphic ( "..\\..\\1CABVIEW\\S100\\ind_MP2.ACE" )
Style ( SPRUNG )
MouseControl ( 0 )
NumFrames ( 2 2 1 )
NumPositions ( 2 0 1  )
NumValues ( 2 0 0.01 )
Orientation ( 1 )
DirIncrease ( 1 )
ScaleRange ( 0 0.01 )
)
```

Some of them are just boolean on/off lights that indicate whether traction and/or dynamic braking is active.

Also, there is another indicator (the one using flecha_intensidadPuff.ace), which works properly for braking, but it shows non-zero throttle when throttle is at zero.

[Textures.zip](https://github.com/openrails/openrails/files/15388801/Textures.zip)

Reply:
I'll test those out, I can see these controls are a bit creative with the use of NumPositions and NumValues compared to what I've been able to test so far.

Reply:
Looks like my consideration to force combined control displays to not show the 'idle' frame when even the slightest amount of dynamic braking was applied should have also been applied when the slightest amount of traction was applied-now the 'idle' frame will only be shown if the combined control is _exactly_ at the split position; that fixes the on/off light.

Unfortunately, the power/braking indicator isn't quite set up ideally to show exactly 0 throttle at idle (you need an odd number of frames for a switch value of exactly 0.5 to be produced, 54 frames produces two values equally spaced [neglecting floating point error] above and below 0.5), so I have added a bit of a bodge that I think will result in the same animation frame selected for 0 throttle as in old versions.

Reply:
I've been told that the indicators are working now, thanks!

Reply:
Good to hear, I realized that my bodge for the idle indication wouldn't work exactly as expected (it would calculate the animation frame of the split position in an MSTS-compatible way, but that wouldn't guarantee that index would actually be used at idle throttle), so it's dumb luck that's working. I know what I missed so I'll get that in later and continue working through a variety of content to look for other things to fix.
It's baffling how MSTS just handled every unusual thing people threw at it.